packages/eslint-plugin/src/rules/strict-enums.ts

```diff
@@ -0,0 +1,720 @@
+       * First, we get all the parts of the union. For non-union types, this
+       * will be an array with the type in it. For example:
+       * - Fruit --> [Fruit]
+       * - Fruit | Vegetable --> [Fruit, Vegetable]
+       */
+      const subTypes = tsutils.unionTypeParts(type);
+
+      /**
+       * Next, we must resolve generic types with constraints. For example:
+       * - Fruit --> Fruit
+       * - T extends Fruit --> Fruit
+       */
+      const subTypesConstraints = subTypes.map(subType => {
+        const constraint = subType.getConstraint();
+        return constraint === undefined ? subType : constraint;
+      });
+
+      const enumSubTypes = subTypesConstraints.filter(subType =>
+        isEnum(subType),
+      );
+      const baseEnumSubTypes = enumSubTypes.map(subType =>
+        getBaseEnumType(subType),
+      );
+      return new Set(baseEnumSubTypes);
+    }
+
+    /**
+     * Given a set A and set B, return a set that contains only elements that are in
+     * both sets.
+     */
+    function getIntersectingSet<T>(a: Set<T>, b: Set<T>): Set<T> {
+      const intersectingValues = [...a.values()].filter(value => b.has(value));
+      return new Set(intersectingValues);
+    }
+
+    /**
+     * From:
+     * https://stackoverflow.com/questions/13627308/add-st-nd-rd-and-th-ordinal-suffix-to-a-number
+     */
+    function getOrdinalSuffix(i: number): string {
+      const j = i % 10;
+      const k = i % 100;
+      if (j == 1 && k != 11) {
+        return `${i}st`;
+      }
+      if (j == 2 && k != 12) {
+        return `${i}nd`;
+      }
+      if (j == 3 && k != 13) {
+        return `${i}rd`;
+      }
+      return `${i}th`;
+    }
+
+    function getTypeFromNode(node: TSESTree.Node): ts.Type {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      return getTypeFromTSNode(tsNode);
+    }
+
+    function getTypeFromTSNode(tsNode: ts.Node): ts.Type {
+      return typeChecker.getTypeAtLocation(tsNode);
+    }
+
+    function getTypeName(type: ts.Type): string {
+      const typeName = util.getTypeName(typeChecker, type);
+      if (
+        typeName.length <= TYPE_NAME_TRUNCATION_THRESHOLD ||
+        insideJestTest() // Never truncate in tests
+      ) {
+        return typeName;
+      }
+
+      const truncatedTypeName = typeName.substring(
+        0,
+        TYPE_NAME_TRUNCATION_THRESHOLD,
+      );
+      return `${truncatedTypeName} [snip]`;
+    }
+
+    function hasEnumTypes(type: ts.Type): boolean {
+      const enumTypes = getEnumTypes(type);
+      return enumTypes.size > 0;
+    }
+
+    function insideJestTest(): boolean {
+      return process.env.JEST_WORKER_ID !== undefined;
+    }
+
+    function isArray(
+      type: ts.Type,
+    ): type is ts.TypeReference | ts.TupleTypeReference {
+      const typeName = getTypeName(type);
+      return (
+        typeChecker.isArrayType(type) ||
+        typeChecker.isTupleType(type) ||
+        typeName.startsWith('Iterable<')
```

Comment:
We also don't compare type names by strings. It's relatively inefficient to make the type name serializer work on them, and you end up with false positives when users define their own types.

I have to go and don't have time to dig into an alternate strategy, but maybe check `no-floating-promises` and `no-for-in-array` to see if they have good things? We generally look at how types behave (`isPromiseLike`) if there isn't a built-in API.

I'm also wondering why it's necessary to check this in the first place? But haven't had time to read through everything yet, sorry.

Reply:
> We also don't compare type names by strings.

Yeah, I know the string comparison is bad, we should replace it with some better code, but I don't know how to do so.

I can't quite follow the code in no-floating-promises, it just looks like its hard-coded to look for a property named "then"? I dont think that helps much here. As for for-in-array, it just uses `checker.isArrayType`, but there is no corresponding `isIterable` method, so I'm stuck.

> I'm also wondering why it's necessary to check this in the first place?

It's for the following test:

```ts
const fruitSet = new Set([
  Fruit.Apple,
  Fruit.Banana,
]);

fruitSet.has(Fruit.Apple);
```

This throws an error that says "Array<Fruit>" does not match "Iterable<Fruit>".
So basically, I need to just treat iterable types the same as an array for the purposes of this rule.

Reply:
> hard-coded to look for a property named "then"? 

Exactly. Instead of checking for whether something _is_ a particular thing, one common strategy is to check whether it _behaves_ like expected. The concept of a Thenable in JavaScript -and therefore TypeScript's types- is roughly an object with a `.then()` that returns a Thenable. Roughly.

> just treat iterable types the same as an array 

I think you're giving yourself a lot harder of a job than you need to. In general, our lint rules don't hardcode cases for types of data structures. Users should never be surprised that, say, built-in `Array` and `Map` work one way but their custom `MyDataStructure` class doesn't.

So, for iterable methods like `.has` and `.includes`, don't give them any special handling. The lint rule probably shouldn't care about `Array<Fruit>` vs. `Iterable<Fruit>` or even know what the container type is. All it should care about is that there's a parameter type, and an argument type, and the matchup shouldn't violate what the rule cares about.

Reply:
Ok. I see now.

I refactored it to use `if (util.isTypeReferenceType(type))`, and then `const typeArguments = typeChecker.getTypeArguments(type)` now, which seems to work.